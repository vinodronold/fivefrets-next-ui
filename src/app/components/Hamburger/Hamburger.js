import React from 'react'
import { bool, func, number } from 'prop-types'
import { color } from '../../constants/theme'

const dim = 5
const _propTypes = {
  n: number.isRequired,
  isOpen: bool.isRequired
}
const _defaultProps = {
  isOpen: false
}

const Layer = ({ n, isOpen }) => (
  <span>
    <style jsx>{`
      span {
        display: block;
        border-radius: 2px;
        height: 4px;
        position: relative;
        margin: 0.5rem 0;
        top: ${dim * 0.24}rem;
        left: ${dim * 0.2}rem;
        background: ${color.primary()};
        transition: 0.25s ease-in-out;
        transition-delay: ${(isOpen ? 0 : 0.5) + 0.125 * (n - 1)}s;
        width: ${isOpen ? 0 : dim * 0.6}rem;
      }
    `}</style>
  </span>
)
Layer.propTypes = _propTypes
Layer.defaultProps = _defaultProps

const Arm = ({ n, isOpen }) => (
  <span>
    <style jsx>{`
      span {
        position: absolute;
        transition: 0.25s ease-in-out;
        background: ${color.primary()};
      }
    `}</style>
    <style jsx>{`
      span {
        top: ${n === 1 ? `30%` : `${dim * 0.48}rem`};
        left: ${n === 1 ? `${dim * 0.48}rem` : `30%`};
        width: ${n === 1 ? `4px` : `${isOpen ? 40 : 0}%`};
        height: ${n === 1 ? `${isOpen ? 40 : 0}%` : `4px`};
        transition-delay: ${n === 1 ? `${isOpen ? 0.625 : 0}s` : `${isOpen ? 0.375 : 0.25}s`};
      }
    `}</style>
  </span>
)
Arm.propTypes = _propTypes
Arm.defaultProps = _defaultProps

const _propTypes_Hamburger = {
  isMenuOpen: bool.isRequired,
  isSearchOpen: bool.isRequired,
  ToggleMenu: func.isRequired,
  ToggleSearch: func.isRequired
}
const _defaultProps_Hamburger = {
  isMenuOpen: false,
  isSearchOpen: false
}
const Hamburger = ({ isMenuOpen, isSearchOpen, ToggleMenu, ToggleSearch }) => (
  <div className="container" onClick={isSearchOpen ? ToggleSearch : ToggleMenu}>
    <div className="layers">
      {Array(3)
        .fill(1)
        .map((_, i) => <Layer key={i} n={i} isOpen={isMenuOpen || isSearchOpen} />)}
    </div>
    <div className="arms">
      {Array(2)
        .fill(1)
        .map((_, i) => <Arm key={i} n={i} isOpen={isMenuOpen || isSearchOpen} />)}
    </div>

    <style jsx>{`
      .container {
        width: ${dim}rem;
        height: ${dim}rem;
        bottom: 1.1rem;
        right: 1.1rem;
        position: fixed;
        cursor: pointer;
        border-radius: 50%;
        z-index: 15;
        transition: all 0.5s;
        background: ${color.secondary()};
      }
      .container:hover {
        background: ${color.secondary(0.8)};
      }
      .layers {
        position: absolute;
        height: 100%;
        width: 100%;
      }
      .arms {
        position: absolute;
        height: 100%;
        width: 100%;
        transform: rotate(45deg);
      }
    `}</style>
  </div>
)
Hamburger.propTypes = _propTypes_Hamburger
Hamburger.defaultProps = _defaultProps_Hamburger

export default Hamburger
export { Layer, Arm }
