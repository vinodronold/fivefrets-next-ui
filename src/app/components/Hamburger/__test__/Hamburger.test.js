import React from 'react'
import renderer from 'react-test-renderer'

import Hamburger, { Layer, Arm } from '../Hamburger'

describe('components/Hamburger.js', () => {

  it('should match Hamburger Snapshot when close', () => {
    const _Hamburger = renderer.create(<Hamburger ToggleMenu={jest.fn()} ToggleSearch={jest.fn()} />)
    expect(_Hamburger).toMatchSnapshot()
  })
  it('should match Hamburger Snapshot when Menu is Open', () => {
    const _Hamburger = renderer.create(
      <Hamburger isMenuOpen={true} ToggleMenu={jest.fn()} ToggleSearch={jest.fn()} />
    )
    expect(_Hamburger).toMatchSnapshot()
  })
  it('should match Hamburger Snapshot when Search is Open', () => {
    const _Hamburger = renderer.create(
      <Hamburger isSearchOpen={true} ToggleMenu={jest.fn()} ToggleSearch={jest.fn()} />
    )
    expect(_Hamburger).toMatchSnapshot()
  })
  it('should match Layer Snapshot when Closed', () => {
    const _Layer = renderer.create(<Layer n={0} />)
    expect(_Layer).toMatchSnapshot()
  })
  it('should match Layer Snapshot when Open', () => {
    const _Layer = renderer.create(<Layer n={0} isOpen={true} />)
    expect(_Layer).toMatchSnapshot()
  })
  it('should match Arm Snapshot when Closed', () => {
    const _Arm = renderer.create(<Arm n={0} />)
    expect(_Arm).toMatchSnapshot()
  })
  it('should match Arm Snapshot when Open', () => {
    const _Arm = renderer.create(<Arm n={0} isOpen={true} />)
    expect(_Arm).toMatchSnapshot()
  })
})
