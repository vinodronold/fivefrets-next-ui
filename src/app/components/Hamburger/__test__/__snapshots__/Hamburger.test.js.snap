// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Hamburger.js should match Arm Snapshot when Closed 1`] = `
<span
  className="jsx-301163176 jsx-179826770"
/>
`;

exports[`components/Hamburger.js should match Arm Snapshot when Open 1`] = `
<span
  className="jsx-301163176 jsx-794770576"
/>
`;

exports[`components/Hamburger.js should match Hamburger Snapshot when Menu is Open 1`] = `
<div
  className="jsx-4125533449 container"
  onClick={[Function]}
>
  <div
    className="jsx-4125533449 layers"
  >
    <span
      className="jsx-1304497301"
    />
    <span
      className="jsx-283106720"
    />
    <span
      className="jsx-3759017912"
    />
  </div>
  <div
    className="jsx-4125533449 arms"
  >
    <span
      className="jsx-301163176 jsx-794770576"
    />
    <span
      className="jsx-301163176 jsx-3128104624"
    />
  </div>
</div>
`;

exports[`components/Hamburger.js should match Hamburger Snapshot when Search is Open 1`] = `
<div
  className="jsx-4125533449 container"
  onClick={[Function]}
>
  <div
    className="jsx-4125533449 layers"
  >
    <span
      className="jsx-1304497301"
    />
    <span
      className="jsx-283106720"
    />
    <span
      className="jsx-3759017912"
    />
  </div>
  <div
    className="jsx-4125533449 arms"
  >
    <span
      className="jsx-301163176 jsx-794770576"
    />
    <span
      className="jsx-301163176 jsx-3128104624"
    />
  </div>
</div>
`;

exports[`components/Hamburger.js should match Hamburger Snapshot when close 1`] = `
<div
  className="jsx-4125533449 container"
  onClick={[Function]}
>
  <div
    className="jsx-4125533449 layers"
  >
    <span
      className="jsx-3221435804"
    />
    <span
      className="jsx-3063298808"
    />
    <span
      className="jsx-3630948284"
    />
  </div>
  <div
    className="jsx-4125533449 arms"
  >
    <span
      className="jsx-301163176 jsx-179826770"
    />
    <span
      className="jsx-301163176 jsx-2834914619"
    />
  </div>
</div>
`;

exports[`components/Hamburger.js should match Layer Snapshot when Closed 1`] = `
<span
  className="jsx-3221435804"
/>
`;

exports[`components/Hamburger.js should match Layer Snapshot when Open 1`] = `
<span
  className="jsx-1304497301"
/>
`;
